### PR TITLE
Move IoError interface from Network to Api namespace

### PR DESCRIPTION
**Pull request by yaelharel**
*Description*:
This PR is broken out from #6072.

*Risk Level*:
Low

*Testing*:
bazel build //source/... && bazel test //test/...

*Docs Changes*:
N/A

*Release Notes*:
N/A

cc @jmarantz @dnoe

### CODE REVIEW

**Comment by mattklein123** [comment]
@danzh2010 PTAL

File: include/envoy/api/io_error.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  IoCallResult(IoCallResult<T>&& result) : rc_(result.rc_), err_(std::move(result.err_)) {}
+
+  virtual ~IoCallResult() {}
+
+  T rc_;
+  IoErrorPtr err_;
+};
+
+using IoCallBoolResult = IoCallResult<bool>;
+using IoCallSizeResult = IoCallResult<ssize_t>;
```
Comment (venilnoronha) [inline include/envoy/api/io_error.h]: It seems that `IoCallBoolResult` and `IoCallSizeResult` types are unused. I'd drop them off.

Comment (yaelharel) [reply include/envoy/api/io_error.h]: They're left overs from #6072. I'll remove them.

File: include/envoy/api/io_error.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+namespace Envoy {
+namespace Api {
+
+class IoError;
+
+// IoErrorCode::Again is used frequently. Define it to be a distinguishable address to avoid
+// frequent memory allocation of IoError instance.
+// If this is used, IoCallResult has to be instantiated with a deleter that does not
+// deallocate memory for this error.
+#define ENVOY_ERROR_AGAIN reinterpret_cast<Api::IoError*>(0x01)
```
Comment (dnoe) [inline include/envoy/api/io_error.h]: I think there was some consensus that this wasn't a good way of doing this. At minimum I'd like to continue this discussion, even though your PR is just moving this. Can you add a TODO comment here, and maybe create an issue for this so we don't lose track?

Comment (yaelharel) [reply include/envoy/api/io_error.h]: Added a TODO in the code with a reference to #6037 since it looks like it is being handled there.

**Pull request merged**